staging/src/k8s.io/client-go/util/workqueue/queuelogger.go

```diff
@@ -0,0 +1,132 @@
+	"time"
+
+	"k8s.io/utils/clock"
+)
+
+func (d DelayingOptions) String() string {
+	var str strings.Builder
+
+	str.WriteString(fmt.Sprintf("Duration:%s/%s", d.Duration, d.WhenWaiting))
+	if d.PermitActiveAndWaiting {
+		str.WriteString(", PermitActiveAndWaiting")
+	}
+	return str.String()
+}
+
+func (w waitFor) String() string {
+	switch w.action {
+	case waitForActionDoneWaiting, waitForActionIsWaiting:
+		return fmt.Sprintf("%s %v", w.action, w.data)
+	case waitForActionLenWaiting, waitForActionNextReady, WaitForActionWaitFill:
+		return w.action.String()
+	}
+	return fmt.Sprintf("%s:%v Ready:%s Options:{%s}", w.action, w.data, format_time(w.readyAt), w.options)
+}
+
+func (w waitForAction) String() string {
+	return [...]string{"Queue", "DoneWaiting", "IsWaiting", "LenWaiting", "NextReady", "WaitFill"}[w]
+}
+
+func (awb AlreadyWaitingBehaviour) String() string {
+	return [...]string{"TakeShorter", "TakeLonger", "TakeIncoming", "TakeExisting"}[awb]
+}
+
+// LogLogfLogger is any logger which supports both Log and Logf (e.g. testing.T)
+type LogLogfLogger interface {
+	Log(args ...any)
+	Logf(format string, args ...any)
+}
+
+// queueLogger is a specialised logger which logs which queue level is logging and adds a timestamp.
+// It has a mutex so that the logger can be used and set from different goroutines.
+type queueLogger struct {
+	name   string
+	logger LogLogfLogger
+	clock  clock.PassiveClock
+}
+
+// NewQueueLogger creates a QueueLogger in the disabled state (i.e. no logger).
+func NewQueueLogger(name string, logger LogLogfLogger, clock clock.PassiveClock) *queueLogger {
+	return &queueLogger{
+		name:   name,
+		logger: logger,
+		clock:  clock,
+	}
+}
+
+// log outputs a message with an optional timestamp if the logger supports it,
+func (l *queueLogger) Log(args ...any) {
+	if l.logger == nil {
+		return
+	}
+	msg := make([]any, 0)
+	// the original logger will be the timekeeper rather than the embedded.
+	if l.clock != nil {
+		msg = append(msg, format_time(l.clock.Now())+":")
+	}
+	msg = append(msg, l.name+":")
+	msg = append(msg, args...)
+	l.logger.Log(msg...)
+}
+
+// logf appends the formatted Unix Time Stamp to the front of the format string.
+func (l *queueLogger) Logf(format string, args ...any) {
+	if l.logger == nil {
+		return
+	}
+	format = l.name + ": " + format
+	if l.clock != nil {
+		format = format_time(l.clock.Now()) + ": " + format
+	}
+	l.logger.Logf(format, args...)
+}
+
+// format_time - makes a time more human parseable by placing a '.' between the seconds and the fractional part.
+func format_time(t time.Time) string {
+	un := t.UnixNano()
+	secs := t.Unix()
+	frac := un - (secs * 1000000000)
+	return strconv.FormatInt(secs, 10) + "." + strconv.FormatInt(frac, 10)
+}
+
+// EnableLogging is a simple way of logging detailed queue events.
+// It recurses into types we know about to set the logging for each of the levels of queue.
+// Call 'workqueue.EnableLogging(t, yourqueuevar)' at the start of your test.
```

Comment:
I agree with @atiratree: please try to avoid maintaining your own logging abstraction, in particular when it gets exposed as part of the public API.

slog might soon become a structured logging API in the Go standard library. I'm investigating whether we can use that as API instead of go-logr. It would be good to wait with changing public APIs in staging repos until the future direction is clear. Internally it's easy to switch between them, but not in a public API.